src/libexpr/eval-inline.hh

```diff
@@ -104,11 +104,16 @@ void EvalState::forceValue(Value & v, Callable getPos)
         }
     }
     else if (v.isApp()) {
-        PosIdx pos = getPos();
-        callFunction(*v.app.left, *v.app.right, v, pos);
+        try {
+            callFunction(*v.app.left, *v.app.right, v, noPos);
+        } catch (InfiniteRecursionError & e) {
+            // only one black hole can *throw* in any given eval stack so we need not
+            // check whether the position is set already.
+            if (v.isBlackhole())
+                e.err.errPos = positions[getPos()];
+            throw;
+        }
     }
-    else if (v.isBlackhole())
-        error("infinite recursion encountered").atPos(getPos()).template debugThrow<EvalError>();
```

Comment:
I'll rebase https://github.com/NixOS/nix/pull/8623 after this :smiling_face_with_tear:

Reply:
While I get the sentiment, I think #8623 will benefit from this PR by moving the only hot path addition out of the hot path.
Most of it happens in `catch` blocks that shouldn't affect performance because they don't run very often at all.